File: Data-Structure-I/383-Ransom-Note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        c = Counter(magazine)
        for i in ransomNote:
            if i in c:
                if c[i] == 1:
                    c.pop(i)
                else:
                    c[i] -= 1
            else:
                return False
        return True

File: Data-Structure-I/383-Ransom-Note.cpp
// solution 1:
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        sort(ransomNote.begin(), ransomNote.end());
        sort(magazine.begin(), magazine.end());
        int i=0, j=0, count=0;
        while(i<ransomNote.length() && j<magazine.length()){
            if(ransomNote[i]==magazine[j]) ++i, ++j, ++count;
            else if(ransomNote[i]<magazine[j]) ++i;
            else ++j;
        }
        return (count==ransomNote.length()) ? true : false;
    }
};

// solution 2:
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        for(int i=0; i<ransomNote.size(); ++i){
            size_t found = magazine.find(ransomNote[i]);
            if(found==-1) return false;
            magazine.erase(found, 1);
        }
    return true;
    }
};

